### DISCUSSION TITLE
command to manually enter DeepSleep and wakeup via serial

### DISCUSSION THREAD

**Discussion by eadmaster** [arendst/Tasmota#21516] [Q&A]
is it possible to set an ESP32 to an unlimited [deepsleep](https://tasmota.github.io/docs/DeepSleep/) (without timer), and then manually wakeup via serial command when the device is needed?

(i'm considering connecting the esp32 to a [NanoPi R1](https://wiki.friendlyelec.com/wiki/index.php/NanoPi_R1) and control it remotely via the serial console)

EDIT: this seems to be pretty close to what i was looking:

1. cmd `Restart 2`  = halt system (needs hardware reset or power cycle to restart)
2. then, to wake up: `esptool --port /dev/ttyUSB0 --before default_reset --after hard_reset chip_id`

Seems to be working fine, haven't noticed any issues with this method so far.
Maybe it will cause unnecessary flash writes if performed frequently?

**Comment by sfromis** [comment]
What you do there has nothing at all to do with "deep sleep", the ESP32 is **not** *sleeping* after `Restart 2`. Of course, that distinction may not be important to you, but you should be able to do the same just using the normal reset pin without involving `esptool`, and the dev board setup it is using to tickle the reset pin via USB. And yes, frequent boots may count against flash lifetime.

**Comment by eadmaster** [comment]
Ok, as alternative, what command could i use to enter deepsleep and wake up manually?

I've found there's `Restart 9` (save all changes and go into deepsleep waiting for a reset), but it seems not available in my build.

**Comment by eadmaster** [comment]
update: i've investigated the esp32 docs a bit more and found it is actually possible to wakeup from **light-sleep** via a serial cmd after setting this:
````
esp_sleep_enable_uart_wakeup(GPIO_NUM_0); // GPIO0 is typically used for UART0 on most ESP32 boards
````

Then you can manually enter light sleep with
````
esp_light_sleep_start();
````

[(source)](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/system/sleep_modes.html)

These commands seems currently not wrapped in tasmota , so i'll try to add them in my test build and see if they work as expected.

I'll also try to make the esp32  enter  (light)sleep mode as soon after booting, so i can turn it on manually only when needed (i do not want to use wifi for remote control and do not want to use a physical switch to wake up).
